modules/ingester/instance.go

```diff
@@ -133,59 +133,77 @@ func (i *instance) CutCompleteTraces(cutoff time.Duration, immediate bool) error
 	return nil
 }
 
-func (i *instance) CutBlockIfReady(maxBlockLifetime time.Duration, maxBlockBytes uint64, immediate bool) error {
+// CutBlockIfReady cuts a completingBlock from the HeadBlock if ready
+// Returns a bool indicating if a block was cut along with the error (if any).
+func (i *instance) CutBlockIfReady(maxBlockLifetime time.Duration, maxBlockBytes uint64, immediate bool) (bool, error) {
 	i.blocksMtx.Lock()
 	defer i.blocksMtx.Unlock()
 
 	if i.headBlock == nil || i.headBlock.DataLength() == 0 {
-		return nil
+		return false, nil
 	}
 
 	now := time.Now()
 	if i.lastBlockCut.Add(maxBlockLifetime).Before(now) || i.headBlock.DataLength() >= maxBlockBytes || immediate {
-		if i.completingBlock != nil {
-			return fmt.Errorf("unable to complete head block for %s b/c there is already a completing block.  Will try again next cycle", i.instanceID)
-		}
+		i.completingBlocks = append(i.completingBlocks, i.headBlock)
 
-		i.completingBlock = i.headBlock
 		err := i.resetHeadBlock()
 		if err != nil {
-			return fmt.Errorf("failed to resetHeadBlock: %w", err)
+			return true, fmt.Errorf("failed to resetHeadBlock: %w", err)
 		}
 
-		// todo : this should be a queue of blocks to complete with workers
-		go func() {
-			completeBlock, err := i.writer.CompleteBlock(i.completingBlock, i)
-			i.blocksMtx.Lock()
-			defer i.blocksMtx.Unlock()
-
-			if err != nil {
-				// this is a really bad error that results in data loss.  most likely due to disk full
-				_ = i.completingBlock.Clear()
-				metricFailedFlushes.Inc()
-				i.completingBlock = nil
-				level.Error(log.Logger).Log("msg", "unable to complete block.  THIS BLOCK WAS LOST", "tenantID", i.instanceID, "err", err)
-				return
-			}
-			i.completingBlock = nil
-			i.completeBlocks = append(i.completeBlocks, completeBlock)
-		}()
+		return true, nil
 	}
 
-	return nil
+	return false, nil
+}
+
+// CompleteBlock() moves a completingBlock to a completeBlock
+func (i *instance) CompleteBlock() bool {
```

Comment:
i think this method needs to complete all blocks in the completingBlocks slice and not just the first one.  the operation you are putting in the queue is "complete blocks for this instance" and so should cover all blocks.

if we changed the operations to identify a specific block then we would only do 1.

Reply:
OK this made me think and uncover a totally new scenario. Consider this case:

- sweepAllInstances(immediate=false) is called from the regular flush cycle
- sweepAllInstances(immediate=true) is called immediate after from shutdownhandler
- the completingBlock from the first call will take a long time to complete (because it will be a large block) and during this time, the `enqueue(opKindComplete)` from the second call will not succeed. 

This is as good as saying "while there's one completingBlock no other can process".

I think we should add blockID to the flushOp. And use that as part of the `Key()` so multiple `opKindComplete`s can be enqueued.

Reply:
this is ok, we just need to make complete behave like flush.  the flush logic keeps flushing the instance until there is nothing left to do.

Reply:
Even if we change it to that, we will miss work in the shutdownhandler right? Consider this:

- complete op is queued and is iterating over the loop of completingBlocks
- it finishes this iteration, checks `len(completingBlocks)` is empty
- before it could quit that loop, "shutdownhandler" is called, this increases length of completingBlocks but the completeOp could not be enqueued because there's already an op in progress. 

That op would be lost

Reply:
you are correct.  this would be lost.  agree we should go to per block operations.